### DISCUSSION TITLE
Trying lexy with AoC and running into errors

### DISCUSSION THREAD

**Discussion by xsebek** [foonathan/lexy#115] [Q&A] [answered]
Hi, I am trying to learn Lexy using Advent of Code, specifically 2017 day 7.

The input has lines like this:
```
ktlj (57)
fwft (72) -> ktlj, cntj, xhth
```

Which I want to parse into a vector of these structures:
```cpp
struct Program {
    std::string name;
    int weight;
    std::vector<std::string> above; // can be empty
}
```

But I run into static asserts that I have trouble understanding. This is my code so far:
```cpp
namespace grammar {
namespace dsl = lexy::dsl;

struct word {
    static constexpr auto rule = dsl::identifier(dsl::ascii::alpha);
    static constexpr auto value = lexy::as_string<std::string>;
};

struct program_list {
    static constexpr auto rule = dsl::list(dsl::inline_<word>, dsl::sep(dsl::comma));
    // dsl::sep ---> static assertion failed due to requirement 'lexy::is_branch_rule<lexyd::_comma>'
    static constexpr auto value = lexy::as_list<std::vector<std::string>>;
};

struct program {
    static constexpr auto rule =
            dsl::p<word> >>
            dsl::whitespace(dsl::round_bracketed(dsl::integer<int>)) >>
            // the rest should be optional, but I did not even get that far
            dsl::whitespace(dsl::lit<"->">) >>
            dsl::whitespace(dsl::inline_<program_list>);
    static constexpr auto value = lexy::construct<Program>;
};
}

auto parse_line(const std::string &input) -> Program {
    auto l_input = lexy::string_input<lexy::ascii_encoding>(input);
    auto result = lexy::parse<grammar::program>(l_input, lexy_ext::report_error);
    // include/lexy/dsl/base.hpp:166:39: error: static assertion failed: pattern rule must not produce any values
    //  166 |         static_assert(sizeof...(Args) == 0, "pattern rule must not produce any values");
    //      |                       ~~~~~~~~~~~~~~~~^~~~
    // include/lexy/grammar.hpp:347:36: error: static assertion failed: missing value callback overload for production
    //  347 |             static_assert(_detail::error<Production, Args...>,
    //      |                           ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
    if (result.has_value()) {
        return result.value();
    }
    throw std::runtime_error("Can not parse program");
}
```

The problem is that I have no idea which part of the parser is causing these issues. The first error with `sep` sounds like the likely culprit, but it is taken almost verbatim from documentation. :confused:

**Comment by foonathan** [comment, marked as answer]
* `// dsl::sep ---> static assertion failed due to requirement 'lexy::is_branch_rule<lexyd::_comma>'`: this error shouldn't happen, `dsl::comma` is a branch rule. 
* `static constexpr auto rule = dsl::list(dsl::inline_<word>, dsl::sep(dsl::comma));`: you want `dsl::p` not `dsl::inline_`. `dsl::inline_` is only meant for contexts like whitespace rules which must not use productions.
* To parse things in sequence, you need `operator+`, not `operator>>`. `operator>>` is only used to differentiate between the condition of a branch, see here: https://lexy.foonathan.net/learn/branching/
* `dsl::whitespace(...)` does not do what you think it does. It parses the inner rule as whitespace, not whitespace than the inner rule. This causes your "pattern rule must not produce any values". If you just define a top-level `whitespace` member, you get whitespace skipping for free: https://lexy.foonathan.net/reference/dsl/whitespace/#doc

Applying those fixes, making the part with `->` optional using `dsl::opt` and using a branch condition, renaming `program_list` to `word_list` and adding a new `program_list` production for parsing the list of programs, the sample input can be parsed like so: https://lexy.foonathan.net/playground/?id=sa8Koxbdh&mode=tree

(The online playground doesn't support value callbacks.)

**Comment by xsebek** [comment]
Thanks a lot, @foonathan! I added `value` callbacks and it works: :partying_face: 
https://godbolt.org/z/9fanv1zh8

Clang error messages on Godbolt were a bit more helpful, but that might just have been luck.

- The `dsl::sep` error was some bogus error shown by the editor, no idea what is confusing it, but it compiles without error.
- Yeah, I used `dsl::p` originally, but then some compile error suggested `inline` and I was desperate. :sweat_smile: 
- Right, so `>>` basically is used to force some conditions on the left side and not backtrack. :thinking: 
- Hah, I must have been thinking about [`Text.Megaparsec.Char.Lexer.space`](https://hackage.haskell.org/package/megaparsec-9.3.0/docs/Text-Megaparsec-Char-Lexer.html#v:space).

**Reply by foonathan** [reply]
> Yeah, I used dsl::p originally, but then some compile error suggested inline and I was desperate

Yes, `dsl::whitespace(dsl::p<T>)` is not allowed, but `dsl::whitespace(dsl::inline_<T>)` is. But you didn't want whitespace there ;)

> Right, so >> basically is used to force some conditions on the left side and not backtrack. 

Yes, lexy will only ever backtrack the LHS of a `>>`.
